### PR TITLE
feat(cdf-tyepscript)!: Add cdk8s and pulumi support in CDF

### PR DESCRIPTION
**Pull request by cwang**
## Description

- Add cdk8s support
- Add pulumi support
- Improve test coverage on runtime specific bootstrap code

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Refactor and code improvement (non-breaking change which improves code quality and/or performance)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation
- [x] Tests
- [ ] Other chores such as maintenance

## How Has This Been Tested?

More dev deps are introduced to establish bootstrap code testing patterns.

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings
- [ ] The commit message follows the convention of this project

### CODE REVIEW

File: core/typescript/src/lib/bootstrap/pulumi/component.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import { Orchestrator } from '../../orchestrator';
+import { InfraPlanConstructs, Planner, ResultHandler } from '../../package';
+import { Custom } from '../../project';
+
+/**
+ * Properties for {@link ProjectComponent}.
+ *
+ * @group For project creators
+ */
+export interface ProjectComponentOpts<
```
Comment (janaka) [inline core/typescript/src/lib/bootstrap/pulumi/component.ts:20]: I feel this should be `Stack` to make the model consistent with the other runtimes. Pulumi does have the concept of a stack though there isn't an a concrete implementation.

image

image

File: core/typescript/src/lib/bootstrap/cdk8s/chart.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import { Orchestrator } from '../../orchestrator';
+import { InfraPlanConstructs, Planner, ResultHandler } from '../../package';
+import { Custom } from '../../project';
+
+/**
+ * Properties for {@link ProjectChart}.
+ *
+ * @group For project creators
+ */
+export interface ProjectChartProps<
```
Comment (janaka) [inline core/typescript/src/lib/bootstrap/cdk8s/chart.ts:21]: I've not spent much time thinking about K8s and not looked closely at CDK8s. Given the word "charts" presume it operates at the service/app level?

If so this only applies to a CDK `Service` type construct. Never a component. does Network and Relation ever apply?

Comment (janaka) [reply core/typescript/src/lib/bootstrap/cdk8s/chart.ts:21]: Being stand alone probably makes sense though. like each microservice repo would have a CDK8s project. Maybe in a macroservice type setup relations would make sense like [Uber domain-oriented microservice architecture](https://www.uber.com/blog/microservice-architecture/). But that's a pretty niche use case I think.

**Review by janaka — COMMENTED** [review]
Left some comments. 

Icons - don't see them referenced anywhere, assume will be used in docs to come?

File: core/typescript/src/lib/bootstrap/pulumi/component.ts (hunk 1 above)
Comment (cwang) [reply core/typescript/src/lib/bootstrap/pulumi/component.ts:20]: `Stack` in Pulumi carries a different meaning I believe, and that's why I chose to use `Component` over `Stack`. Same goes for cdk8s where `Chart` is used.

File: core/typescript/src/lib/bootstrap/cdk8s/chart.ts (hunk 2 above)
Comment (cwang) [reply core/typescript/src/lib/bootstrap/cdk8s/chart.ts:21]: Good point. I think right now it is probably how it works, i.e. one cdk8s chart => one k8s chart.

Comment (cwang) [reply core/typescript/src/lib/bootstrap/cdk8s/chart.ts:21]: > Left some comments.
> 
> Icons - don't see them referenced anywhere, assume will be used in docs to come?

When we do the release I will manually update that release PR to include those icons

File: core/typescript/src/lib/bootstrap/pulumi/component.ts (hunk 1 above)
Comment (janaka) [reply core/typescript/src/lib/bootstrap/pulumi/component.ts:20]: makes sense in for cdk8s makes sense.

what's the difference in meaning of `stack` carried by Pulumi? I see there's a difference in the implementation. I might be misunderstanding.

Comment (cwang) [reply core/typescript/src/lib/bootstrap/pulumi/component.ts:20]: https://www.pulumi.com/docs/intro/concepts/stack/ which sounds like env in your language or stage in my language :)

Comment (janaka) [reply core/typescript/src/lib/bootstrap/pulumi/component.ts:20]: Reading a little more about `stack` in CDK and Pulumi worlds. As concepts they sounds identical. Implementation differs a little. A concrete implementation of stack does exist in Pulumi also, it's just doesn't need to be explicitly created/referenced in code unless you want to. 

https://docs.aws.amazon.com/cdk/v2/guide/stacks.html

https://www.pulumi.com/learn/building-with-pulumi/understanding-stacks/

Comment (cwang) [reply core/typescript/src/lib/bootstrap/pulumi/component.ts:20]: Not sure I agree. 

- For AWS CDK, you can have multiple stacks within a single account/env, which is also how cloudformation works (stacks and sub-stacks are CF concepts too).
- For Pulumi, stacks == envs and the example they give seem to confirm that too.

File: core/typescript/package.json [hunk 3]
```diff
@@ -63,10 +63,14 @@
...
     "cdktf": "0.x",
-    "@cdktf/provider-google": "^4.0.3",
+    "cdk8s": "2.x",
+    "@pulumi/pulumi": "3.x",
     "constructs": "10.x"
   },
   "devDependencies": {
+    "cdk8s-plus-25": "^2.2.1",
+    "@cdktf/provider-google": "^4.0.3",
+    "@pulumi/azure-native": "^1.0.0",
```
Comment (cwang) [inline core/typescript/package.json:73]: Moving them to dev deps

File: core/typescript/src/lib/metadata-minimal.spec.json [hunk 4]
```diff
@@ -31,9 +30,6 @@
         "hello-world": { "start": "world", "finish": "hello", "bidi": true },
         "fizz-buzz": {"start": "fizz", "finish": "buzz", "bidi": false }
       }
-    },
-    "custom": {
-      "template": "template.ts"
```
Comment (cwang) [inline core/typescript/src/lib/metadata-minimal.spec.json:36]: See below for why `custom` and `network` become optional

File: core/typescript/src/lib/metadata-normal.spec.json [hunk 5]
```diff
@@ -26,19 +26,19 @@
...
       "types": {
-        "world": { "type": "solar-system", "subtype": "earth", "deployment": "NASA would help" },
+        "world": { "type": "solar-system", "subtype": "earth", "deployment": "NASA would help", "label": "Planet Earth" },
         "buzz": { "type": "country", "subtype": "morocco", "deployment": "Fly over", "icon": "https://abc.com/morocco.png", "description": "Well done in World Cup" }
       }
     },
     "relations": {
       "types": {
-        "hello-world": { "start": "world", "finish": "hello", "bidi": true },
-        "fizz-buzz": {"start": "fizz", "finish": "buzz", "bidi": false }
```
Comment (cwang) [inline core/typescript/src/lib/metadata-normal.spec.json:41]: See below for why `label` becomes a new optional field

File: core/typescript/src/lib/metadata.ts [hunk 6]
```diff
@@ -17,6 +17,14 @@ export interface OptionalIconAware {
...
 }
 
+/**
+ * Interface that offers an optional label field.
+ * This should be short and concise.
+ */
+export interface OptionalLabelAware {
+  readonly label?: string;
+}
+
```
Comment (cwang) [inline core/typescript/src/lib/metadata.ts:27]: This is mostly for diagram tooling, from the latest discussion around how to build it where each component|service|relation type needs a label, i.e. name to show on a diagram, alongside description and icon which are already here.

File: core/typescript/src/lib/metadata.ts [hunk 7]
```diff
@@ -99,6 +108,17 @@ export interface ServiceType
...
+ * Interface that represents a typing classification for relations.
+ *
+ * @group For both project creators and package authors
+ */
+export interface RelationType
+  extends Connectable,
+    OptionalLabelAware,
+    OptionalIconAware,
+    OptionalDescAware {}
+
```
Comment (cwang) [inline core/typescript/src/lib/metadata.ts:121]: This is again from the latest discussion around diagram tooling where each line connecting a service/component with another service/company needs to have more info such as label/description/icon similarly to how it's been set up with each component/service type.

File: core/typescript/src/lib/metadata.ts [hunk 8]
```diff
@@ -185,9 +206,10 @@ export interface ConstructsInfo {
    */
   readonly relations: RelationsInfo;
   /**
-   * Information about custom (code blocks).
+   * Information about optional custom (code blocks).
+   * Missing this field implies that the package does NOT support custom code blocks.
    */
-  readonly custom: CustomInfo;
+  readonly custom?: CustomInfo;
```
Comment (cwang) [inline core/typescript/src/lib/metadata.ts:212]: Custom code block as a feature should be optional as a way to limit capabilities offered by a package. A case in point is to allow only service/component customisation in k8s using cdk8s

File: core/typescript/src/lib/metadata.ts [hunk 9]
```diff
@@ -169,9 +189,10 @@ export interface ConstructsInfo {
    */
   readonly schema: string;
   /**
-   * Information about network.
+   * Information about optional network.
+   * Missing this field implies that the package does NOT support customisable virtual network.
    */
-  readonly network: NetworkInfo;
+  readonly network?: NetworkInfo;
```
Comment (cwang) [inline core/typescript/src/lib/metadata.ts:195]: VPC in some public cloud providers could be out of reach, or a package could choose to disallow it

**Review by cwang — COMMENTED** [review]
Customised manifest format a little out of feedback from building diagram tool

File: core/typescript/src/lib/bootstrap/pulumi/component.ts (hunk 1 above)
Comment (janaka) [reply core/typescript/src/lib/bootstrap/pulumi/component.ts:20]: OK I think I get what you are saying. 

Pulumi - a project has multiple instances of a stack (same shape different config like dev and prod).

CDK - an `App` can have multiple stacks of different shape. So with CDK you could have a front end stack and backend stack in the same program. can deploy multiple instances for env but this isn't hard enforced as opposed to Pulumi. 

Is that^ how you are seeing it?

I wonder how Pulumi map this with their CDK implementation https://www.pulumi.com/blog/aws-cdk-on-pulumi/

I'll run the example to see what it does.

My main concern here is that we are putting Pulumi component and CDK stack at the same level which is pretty jarring.

Comment (cwang) [reply core/typescript/src/lib/bootstrap/pulumi/component.ts:20]: In CDK `stack` is exclusively used for a unit of work, much like what a component means with Pulumi. It's also used in CloudFormations as a term. I'm guessing CDK stack == Pulumi component is how Pulumi sees it. 
I actually think it's okay if not a true reflection of the underlying IaC runtime varieties. Bear in mind that the code you see here already exists in a folder such as `bootstrap/pulumi` which is supposed to branch out in individual runtimes, therefore it makes more sense to respect and reuse the terms used in them to describe top-level construct accordingly.

Comment (cwang) [reply core/typescript/src/lib/bootstrap/pulumi/component.ts:20]: These are helper code for ease of integration for a specific runtime and visible within a project codebase. The consequence of using a common term here is that people will see this in their project that conforms to a specific runtime which carries its own set of terms. If we want to be term-compliant, we should start from here.

Comment (janaka) [reply core/typescript/src/lib/bootstrap/pulumi/component.ts:20]: Maybe the plural `*Components` will help with the metal model and still map to pulumi concepts.

**Review by janaka — APPROVED** [review]

**Pull request merged**